spores-pickling/src/main/scala/scala/spores/SporePicklers.scala

```diff
@@ -0,0 +1,372 @@
+  type FullU2[T1, T2, R] = Unpickler[Spore2[T1, T2, R]]
+
+  type FullP3[T1, T2, T3, R] = Pickler[Spore3[T1, T2, T3, R]]
+  type FullU3[T1, T2, T3, R] = Unpickler[Spore3[T1, T2, T3, R]]
+
+  /* Unify type for returning both a pickler and unpickler
+   * that keeps track of the captured type inside a `SporeWithEnv` */
+  type FullPE[T, R, C] = Pickler[SporeWithEnv[T, R] {type Captured = C}]
+  type FullUE[T, R, C] = Unpickler[SporeWithEnv[T, R] {type Captured = C}]
+  type FullPUE[T, R, C] = FullPE[T, R, C] with FullUE[T, R, C]
+
+  type FullPE2[T1, T2, R, C] = Pickler[Spore2WithEnv[T1, T2, R] {type Captured = C}]
+  type FullUE2[T1, T2, R, C] = Unpickler[Spore2WithEnv[T1, T2, R] {type Captured = C}]
+  type FullPUE2[T1, T2, R, C] = FullPE2[T1, T2, R, C] with FullUE2[T1, T2, R, C]
+
+  type FullPE3[T1, T2, T3, R, C] = Pickler[Spore3WithEnv[T1, T2, T3, R] {type Captured = C}]
+  type FullUE3[T1, T2, T3, R, C] = Unpickler[Spore3WithEnv[T1, T2, T3, R] {type Captured = C}]
+  type FullPUE3[T1, T2, T3, R, C] = FullPE3[T1, T2, T3, R, C] with FullUE3[T1, T2, T3, R, C]
+
+  /********************************* Picklers *********************************/
+
+  /* `SporeWithEnv` has their own picklers and unpicklers and these unpicklers
+   * can be picked in the implicit search if we don't look them up for the
+   * subclass `Spore`. Otherwise, a `SporeWithEnv` can be disguised as a
+   * `Spore` and in those cases we use the general `Unpickler` defined below.
+   *
+   * Note that we could ask for the implicits of pickler and unpickler of `U`
+   * here but we don't. The reason is that the scala compiler will interpret
+   * the type `U` to be `this.Captured` instead of the dealiased type. This
+   * spoils the type signatures and makes scala pickling to generate new
+   * picklers for `this.Captured` every time, slowing down the compilation time.
+   */
+
+  implicit def genSporePicklerUnpickler[T, R, U]: FullPUE[T, R, U] =
+      macro genSporePicklerUnpicklerImpl[T, R, U]
+
+  implicit def genSpore2PicklerUnpickler[T1, T2, R, U]: FullPUE2[T1, T2, R, U] =
+      macro genSpore2PicklerUnpicklerImpl[T1, T2, R, U]
+
+  implicit def genSpore3PicklerUnpickler[T1, T2, T3, R, U]: FullPUE3[T1, T2, T3, R, U] =
+      macro genSpore3PicklerUnpicklerImpl[T1, T2, T3, R, U]
+
+  /* Don't make these Picklers to have an `Unpickler` in the return type
+   * since this will screw up the implicit search for any given `Spore` */
+
+  implicit def genSimpleSporePickler[T, R]: FullP[T, R] =
+    macro genSimpleSporePicklerImpl[T, R]
+
+  implicit def genSimpleSpore2Pickler[T1, T2, R]: FullP2[T1, T2, R] =
+    macro genSimpleSpore2PicklerImpl[T1, T2, R]
+
+  implicit def genSimpleSpore3Pickler[T1, T2, T3, R]: FullP3[T1, T2, T3, R] =
+      macro genSimpleSpore3PicklerImpl[T1, T2, T3, R]
+
+  /******************************** Unpicklers ********************************/
+
+  /* These `Unpickler`s are meant to serialize both `Spore`s and `SporeWitEnv`s.
```

Comment:
You mean _deserialize_ here, right? `FullU` is only an unpickler as far as I can see. Also, why not use the term "unpickle" instead of "deserialize", since that is more consistent with our usual terminology. ;-)